### PR TITLE
On type mismatch involving fn/method call, point at definition

### LINKED ISSUE #61066: On type mismatch involving fn call, point at fn definition
[Given](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2018&gist=27a5f6a2fb95cd907f8e5511736c82bf)

```
fn foo() -> u32 {
    1
}

fn bar() {
    let x: usize = foo();
}
```

we now present:
```
error[E0308]: mismatched types
 --> src/lib.rs:6:20
  |
6 |     let x: usize = foo();
  |                    ^^^^^ expected usize, found u32
help: you can convert an `u32` to `usize` and panic if the converted value wouldn't fit
  |
6 |     let x: usize = foo().try_into().unwrap();
  |                    ^^^^^^^^^^^^^^^^^^^^^^^^^
```

We should ideally also point at `foo` and maybe tell the user to consider changing it.

### PR DESCRIPTION
**Pull request by sjwang05**
Closes #61066

### CODE REVIEW

File: tests/ui/associated-type-bounds/elision.stderr [hunk 1]
```diff
@@ -21,6 +21,8 @@ LL | fn f(x: &mut dyn Iterator<Item: Iterator<Item = &'_ ()>>) -> Option<&'_ ()>
    |
    = note: expected enum `Option<&()>`
               found enum `Option<impl Iterator<Item = &'_ ()>>`
+note: the method next is defined here
+  --> $SRC_DIR/core/src/iter/traits/iterator.rs:LL:COL
```
Comment (sjwang05) [inline tests/ui/associated-type-bounds/elision.stderr:25]: Not sure what's up with cases like this one--FWIW the span gets printed correctly when I run locally.

File: compiler/rustc_hir_typeck/src/demand.rs [hunk 2]
```diff
@@ -1170,6 +1171,56 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
...
+                kind: hir::TraitItemKind::Fn(sig, ..),
+                ..
+            })
+            | hir::Node::ImplItem(hir::ImplItem {
+                kind: hir::ImplItemKind::Fn(sig, ..), ..
+            })
+            | hir::Node::Item(hir::Item { kind: hir::ItemKind::Fn(sig, ..), .. }) = node
+            && let ret_span = sig.decl.output.span()
+            && !ret_span.from_expansion()
+            && expected.has_concrete_skeleton()
```
Comment (compiler-errors) [inline compiler/rustc_hir_typeck/src/demand.rs:1212]: Why is this needed?

File: compiler/rustc_hir_typeck/src/demand.rs [hunk 3]
```diff
@@ -1170,6 +1171,56 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
...
+        err.span_note(
+            self.tcx.def_span(def_id),
+            format!("the {callee_str} {ident} is defined here"),
+        );
+
+        if let Some(local_did) = def_id.as_local()
+            && let Some(node) = self.tcx.opt_hir_node(self.tcx.local_def_id_to_hir_id(local_did))
+            && let hir::Node::TraitItem(hir::TraitItem {
+                kind: hir::TraitItemKind::Fn(sig, ..),
+                ..
```
Comment (compiler-errors) [inline compiler/rustc_hir_typeck/src/demand.rs:1204]: I don't think it makes sense to suggest changing the definition of a trait. Trait signatures are often the source of truth, and changing a trait method will often break *other* implementations in the crate.

**Review by compiler-errors — CHANGES_REQUESTED** [review]

File: compiler/rustc_hir_typeck/src/demand.rs (hunk 2 above)
Comment (sjwang05) [reply compiler/rustc_hir_typeck/src/demand.rs:1212]: `!from_expansion` is there mainly so we don't suggest changing the return type for desugarings like `async fn`, and `has_concrete_skeleton` is there so we don't suggest things like `_`, or, in cases like the following, we don't suggest changing foo's return type to `T`, which would be nonsensical:

```rs
fn foo() -> i32 {
    42
}

fn bar<T>() -> T {
    foo()
}
```

File: compiler/rustc_hir_typeck/src/demand.rs (hunk 3 above)
Comment (sjwang05) [reply compiler/rustc_hir_typeck/src/demand.rs]: Changed so we only suggest on `Item` and `ImplItem`s, thanks.

File: compiler/rustc_hir_typeck/src/demand.rs [hunk 4]
```diff
@@ -1170,6 +1171,52 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
...
+        {
+            (def_id, path.segments[0].ident, "function")
+        } else if let hir::ExprKind::MethodCall(method, ..) = expr.kind
+            && let Some(def_id) = self.typeck_results.borrow().type_dependent_def_id(expr.hir_id)
+            && !matches!(self.tcx.def_kind(def_id), hir::def::DefKind::Ctor(..))
+        {
+            (def_id, method.ident, "method")
+        } else {
+            return;
+        };
```
Comment (cjgillot) [inline compiler/rustc_hir_typeck/src/demand.rs:1194]: Should we use tcx.def_descr for callee_str?

File: compiler/rustc_hir_typeck/src/demand.rs [hunk 5]
```diff
@@ -1170,6 +1171,52 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
...
+            && let hir::Node::ImplItem(hir::ImplItem {
+                kind: hir::ImplItemKind::Fn(sig, ..), ..
+            })
+            | hir::Node::Item(hir::Item { kind: hir::ItemKind::Fn(sig, ..), .. }) = node
+            && let ret_span = sig.decl.output.span()
+            && !ret_span.from_expansion()
+            && expected.has_concrete_skeleton()
+        {
+            let sugg =
+                if ret_span.is_empty() { format!("-> {expected}") } else { format!("{expected}") };
```
Comment (cjgillot) [inline compiler/rustc_hir_typeck/src/demand.rs:1211]: This should match on `sig.decl.output`.

File: compiler/rustc_hir_typeck/src/demand.rs [hunk 6]
```diff
@@ -1170,6 +1171,52 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
...
+        );
+
+        if let Some(local_did) = def_id.as_local()
+            && let Some(node) = self.tcx.opt_hir_node(self.tcx.local_def_id_to_hir_id(local_did))
+            && let hir::Node::ImplItem(hir::ImplItem {
+                kind: hir::ImplItemKind::Fn(sig, ..), ..
+            })
+            | hir::Node::Item(hir::Item { kind: hir::ItemKind::Fn(sig, ..), .. }) = node
+            && let ret_span = sig.decl.output.span()
+            && !ret_span.from_expansion()
```
Comment (cjgillot) [inline compiler/rustc_hir_typeck/src/demand.rs:1207]: Why not from expansions?

File: compiler/rustc_hir_typeck/src/demand.rs [hunk 7]
```diff
@@ -1170,6 +1171,52 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
...
+        };
+        err.span_note(
+            self.tcx.def_span(def_id),
+            format!("the {callee_str} {ident} is defined here"),
+        );
+
+        if let Some(local_did) = def_id.as_local()
+            && let Some(node) = self.tcx.opt_hir_node(self.tcx.local_def_id_to_hir_id(local_did))
+            && let hir::Node::ImplItem(hir::ImplItem {
+                kind: hir::ImplItemKind::Fn(sig, ..), ..
```
Comment (cjgillot) [inline compiler/rustc_hir_typeck/src/demand.rs:1203]: How do we check we aren't suggesting a change in a trait impl?

File: compiler/rustc_hir_typeck/src/demand.rs [hunk 8]
```diff
@@ -1170,6 +1171,52 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
...
+    ) {
+        let (def_id, ident, callee_str) = if let hir::ExprKind::Call(fun, _) = expr.kind
+            && let hir::ExprKind::Path(hir::QPath::Resolved(_, path)) = fun.kind
+            && let hir::def::Res::Def(def_kind, def_id) = path.res
+            && !matches!(def_kind, hir::def::DefKind::Ctor(..))
+        {
+            (def_id, path.segments[0].ident, "function")
+        } else if let hir::ExprKind::MethodCall(method, ..) = expr.kind
+            && let Some(def_id) = self.typeck_results.borrow().type_dependent_def_id(expr.hir_id)
+            && !matches!(self.tcx.def_kind(def_id), hir::def::DefKind::Ctor(..))
```
Comment (cjgillot) [inline compiler/rustc_hir_typeck/src/demand.rs:1189]: This last condition is common between both chains.
Should it be put a few lines below, with a set of accepted DefKinds instead of refusing Ctor?
I mean some `let DefKind::Fn | DefKind::AssocFn = tcx.def_kind(def_id) else { return };`

File: compiler/rustc_hir_typeck/src/demand.rs [hunk 9]
```diff
@@ -1170,6 +1171,52 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
...
+            self.tcx.def_span(def_id),
+            format!("the {callee_str} {ident} is defined here"),
+        );
+
+        if let Some(local_did) = def_id.as_local()
+            && let Some(node) = self.tcx.opt_hir_node(self.tcx.local_def_id_to_hir_id(local_did))
+            && let hir::Node::ImplItem(hir::ImplItem {
+                kind: hir::ImplItemKind::Fn(sig, ..), ..
+            })
+            | hir::Node::Item(hir::Item { kind: hir::ItemKind::Fn(sig, ..), .. }) = node
```
Comment (cjgillot) [inline compiler/rustc_hir_typeck/src/demand.rs:1205]:
```suggestion
            && let Some(sig) = node.fn_sig()
```

File: compiler/rustc_hir_typeck/src/demand.rs (hunk 6 above)
Comment (sjwang05) [reply compiler/rustc_hir_typeck/src/demand.rs:1207]: Mainly to avoid giving weird suggestions in desugaring like `async fn`, though if you think we should still give suggestions in macros, we could check for `desugaring_kind().is_none()` instead.

File: compiler/rustc_hir_typeck/src/demand.rs (hunk 7 above)
Comment (sjwang05) [reply compiler/rustc_hir_typeck/src/demand.rs:1203]: We don't match on `hir::TraitItem`--I made the check a bit clearer when addressing https://github.com/rust-lang/rust/pull/119340#discussion_r1438398926

File: compiler/rustc_hir_typeck/src/demand.rs (hunk 8 above)
Comment (sjwang05) [reply compiler/rustc_hir_typeck/src/demand.rs]: Added a separate `if` to match on `DefKind::AssocFn` and `DefKind::Fn`, thanks.

**Comment by sjwang05** [comment]
Thanks for all the suggestions.

@rustbot review

**Review by compiler-errors — CHANGES_REQUESTED** [review]
Sorry for any unintended harshness, but I am really not in favor of this change 😅

I believe it inverts the normal "source of truth" we use (the signature) when suggesting to change a function signature that may have other call-sites, and I am skeptical that adding this information on *every* mismatch is worthwhile. Does it account for the other call-sites of the function that may break when this signature is changed?

As for the technical impl, I believe it needs to avoid labeling spans that don't actually map to lines in the source map, and needs to use `backticks` on item names.

Also, does this handle polymorphic functions correctly? How does it map a type mismatch back to the generic parameters of the function, for example?

Finally, why do we need to provide a structured suggestion -- seems to break `run-rustfix` tests for a very heuristic suggestion?

**Comment by Dylan-DPC** [comment]
Closing this as inactive. Feel free to reöpen this pr or create a new pr if you get the time to work on this. Thanks
